### PR TITLE
fix(safety): exclude PDO method calls from exec() false-positive detection

### LINKED ISSUE #19: [Bug] PHP safety scan flags PDO::exec() as dynamic code execution
**Describe the bug**

The PHP safety scanner reports legitimate `PDO::exec()` method calls as `exec()` dynamic-code-execution findings. This creates false positives in database infrastructure, migrations, rebuild tooling, and tests.

**To Reproduce**

1. Run `codexa quality --path . --safety-only --pipe` against a PHP project using PDO.
2. Observe warnings such as `exec() call — avoid dynamic code execution` for calls like `\$connection->exec(...)` and `\$pdo->exec(...)`.
3. In this reproduction, raw search confirms the matches are PDO calls in `app/Core/Database/ConnectionFactory.php`, `app/Core/Database/Migrator.php`, `database/rebuild.php`, and test fixtures.

**Expected behavior**

The safety rule should distinguish the global PHP `exec()` function from method calls such as `PDO::exec()` / `\$pdo->exec()`. PDO SQL execution should not be classified as dynamic shell execution by this rule.

**Environment**

- OS: Linux
- Python: 3.14.7
- CodexA: 0.5.0
- Parser: PHP tree-sitter available

**Additional context**

The same scan also reports other findings, but the PDO `exec()` results are clearly syntactic false positives. An AST-aware or token-aware rule that checks the call target would avoid this noise while retaining detection of actual global `exec()` calls.

### PR DESCRIPTION
**Pull request by elhussienysabry**
## Summary

Fixes #19

The PHP safety scanner was reporting **false positives** for legitimate PDO method calls like `$pdo->exec()` and `PDO::exec()`, flagging them as _dynamic code execution_ findings.

## Root Cause

The `exec()` danger pattern used a simple word-boundary regex:
```python
(r"\bexec\s*\(", "exec() call — avoid dynamic code execution")
```
This matched **any** `exec(` — including PHP PDO method calls like `$connection->exec($sql)`.

## Fix

Added **negative lookbehinds** for `->` and `::` so only standalone function calls are flagged:
```python
(r"(?<!->)(?<!::)\bexec\s*\(", "exec() call — avoid dynamic code execution")
```

| Pattern | Before | After |
|---|---|---|
| `exec('cmd')` | ⛔ flagged | ⛔ flagged (correct) |
| `exec(user_code)` | ⛔ flagged | ⛔ flagged (correct) |
| `$pdo->exec($sql)` | ❌ false positive | ✅ allowed |
| `$connection->exec($sql)` | ❌ false positive | ✅ allowed |
| `PDO::exec($stmt)` | ❌ false positive | ✅ allowed |

## Files Changed
- `semantic_code_intelligence/llm/safety.py` — 1-line regex fix
- `semantic_code_intelligence/tests/test_phase12.py` — 6 regression tests

### CODE REVIEW

**Review by M9nx — COMMENTED** [review]
Request changes

**Inline comment on `semantic_code_intelligence/llm/safety.py` line 26:**

This lookbehind only excludes `->exec` and `::exec` when the operator is immediately adjacent to `exec`. PHP permits whitespace around these operators, so valid calls such as `$pdo -> exec($sql)` and `SomeClass :: exec($sql)` are still reported as dynamic code execution. Please make the matching token-aware or normalize whitespace around `->`/`::` before applying the rule, and add regression tests for both whitespace-separated forms.

The current tests cover the compact forms and pass, but they do not cover this valid PHP formatting case.

File: semantic_code_intelligence/llm/safety.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
     (r"\brm\s+-rf\s+/", "Destructive rm -rf / command"),
     # Dynamic code execution
     (r"\beval\s*\(", "eval() call — avoid dynamic code execution"),
-    (r"\bexec\s*\(", "exec() call — avoid dynamic code execution"),
+    (r"(?<!->)(?<!::)\bexec\s*\(", "exec() call — avoid dynamic code execution"),
```
Comment (M9nx) [inline semantic_code_intelligence/llm/safety.py:30]: This still flags valid PHP when whitespace surrounds the call operator, for example `$pdo -> exec($sql)` or `SomeClass :: exec($sql)`, because the lookbehinds only match operators immediately adjacent to `exec`. Please normalize whitespace or use token-aware matching, and add regression tests for these forms.

**Comment by elhussienysabry** [comment]
@M9nx This PR fixes issue #19 — the PHP safety scanner was generating false positives for legitimate PDO method calls like `$pdo->exec()` and `PDO::exec()`, incorrectly flagging them as dynamic code execution.

**Fix is minimal and safe:**
- 1-line regex change in `semantic_code_intelligence/llm/safety.py`
- Added 6 regression tests confirming:
  - ✅ Global `exec()` is still correctly flagged
  - ✅ `$pdo->exec()`, `$connection->exec()`, `PDO::exec()` are no longer false positives

Ready for review and merge. 🙏

**Comment by M9nx** [comment]
@elhussienysabry Following up on your comment: the compact cases are fixed, but the current regex still flags valid PHP when whitespace surrounds the call operator.

POC against commit `85f7f3c`:

```python
from semantic_code_intelligence.llm.safety import SafetyValidator

validator = SafetyValidator()
for code in ["$pdo -> exec($sql)", "PDO :: exec($sql)"]:
    print(code, validator.is_safe(code))
```

Output:

```text
$pdo -> exec($sql) False
PDO :: exec($sql) False
```

These are still false positives because the lookbehinds only match `->` and `::` when immediately adjacent to `exec`. Please add regression tests for the whitespace-separated forms and update the matcher. This is the same concern raised in the inline review conversation: https://github.com/M9nx/CodexA/pull/28#discussion_r3948818403

**Comment by elhussienysabry** [comment]
Thanks for the review @M9nx! 

You're absolutely right — the fixed-width lookbehinds didn't handle whitespace around `->` and `::`.

**Fix applied in the latest commit:**

Instead of trying to extend the lookbehinds (which require fixed width in Python's `re` module), I added a pre-processing step in `SafetyValidator.validate()` that normalises whitespace around `->` and `::` **before** any pattern is applied:

```python
_PHP_OPERATOR_WS = re.compile(r'\s*(->|::)\s*')

# In validate(), per line:
normalised = _PHP_OPERATOR_WS.sub(r'\1', line)
```

This collapses `$pdo -> exec($sql)` → `$pdo->exec($sql)` and `SomeClass :: exec($sql)` → `SomeClass::exec($sql)` so the existing lookbehinds work for **all** valid PHP spacing styles.

**New regression tests added for spaced forms:**
- `$pdo -> exec($sql)` ✅ not flagged  
- `$connection -> exec($sql)` ✅ not flagged  
- `PDO :: exec($stmt)` ✅ not flagged  
- `SomeClass :: exec($sql)` ✅ not flagged  
- `result = exec(user_input)` ⛔ still flagged  

All 9 test cases pass. Ready for re-review! 🙏

**Review by M9nx — APPROVED** [review]
